### ISSUE TITLE
Chp.1 Using neural nets to recognize handwritten digits

### ISSUE THREAD

**Issue by hysic** [hysic/NNDL_ReadingNotes#1]
## 术语卡

* 术语：Peceptron（感知机）
* 印象：
    * Frank Rosenblatt在1957年提出
    * 其规则为（其中`x`是输入，`w`是权重（weights），`b`是阈值（threshold）。输出为二元输出。
）：
![image](https://cloud.githubusercontent.com/assets/5415646/21916843/29dae22e-d97e-11e6-8fef-4aead90a1b77.png)
    * 利用感知机网络可以实现任意逻辑函数（如`NAND`与非），而`NAND`对于数学计算是universal的，因此感知机对于数学计算也是universal的。比如用感知机实现按位求和（bitwise）（其中每个圆圈代表相同的感知机（w_1 = w_2 = -2, b = 3））：
![image](https://cloud.githubusercontent.com/assets/5415646/21918692/05151480-d98f-11e6-89ee-25a0fdfb2cad.png)
    * 值得一提的是，我们常听见这样的说法：[感知机无法解决异或问题](https://en.wikipedia.org/wiki/Perceptron)，或者更一般的说法，感知机无法解决线性不可分的问题。这里的感知机指的是单层感知机，Marvin Minsky提出这一观点是在1969年，那时貌似还没有多层感知机的BP算法或核方法（待补充）。这与上一条所说的感知机对数学计算是universal的并不矛盾，那里指的是多层感知机。
        * 单隐层（2个hidden neurons）即可解决异或（XOR）问题，见下图
![image](https://user-images.githubusercontent.com/5415646/27823351-664563c6-60db-11e7-8cab-25c098a5a3d9.png)

* 出处
    * [perceptron wiki](https://en.wikipedia.org/wiki/Perceptron)
    * <http://computing.dcu.ie/~humphrys/Notes/Neural/single.neural.html>

**Comment by hysic** [comment]
如何让感知机网络具有学习能力？
-> 不断修改权重和bias，让输出结果向真实结果靠近
-> 感知机无法做到这一点，因为阶梯函数不连续，权重或bias的很小改变可能导致输出发生很大变化
-> sigmoid神经元可以做到这一点，这是一个连续函数，可以返回(0, 1)上的任意值，在x绝对值较大时与感知机函数类似，其图像为：
![image](https://cloud.githubusercontent.com/assets/5415646/22862668/312edff0-f16e-11e6-83ac-cc5082224a1e.png)

**Comment by hysic** [comment]
## 术语卡
* 术语：前馈神经网络（feedforward neural network）
* 印象：网络中没有循环（区别于[RNN（递归神经网络）](http://en.wikipedia.org/wiki/Recurrent_neural_network)），信息只向前传递，不会向后传递
* 出处：[wiki](https://en.wikipedia.org/wiki/Feedforward_neural_network)

**Comment by hysic** [comment]
* 手写数字分类问题 =  segmentation problem + classifying individual digits
* segmentation问题解决方法之一：固定分类器，对手写数字图像进行多次分割，如果分类器对某次分割的得分很高，则说明该分割很好；反之则说明分割很差。

---

* 输出层选择10个神经元，分别对应于10个数字，而不是选择4个神经元，以二进制方式对应于0-9（最多可对应于16个数字），原因是二者相比，前者的性能更好。

**Comment by hysic** [comment]
## 代价函数（Cost function）

![image](https://cloud.gi
thubusercontent.com/assets/5415646/25610074/65b5ae56-2f54-11e7-8193-9a4dcba905df.png)

* Q1：为什么要将代价函数最小化，而不是将正确分类的图片个数（精度）最大化？ -> 因为后者不是光滑函数，w和b发生微小变化不会对后者造成影响，这样我们就无法判断变化方向是否正确。而代价函数是光滑函数，可以很清晰地看到w和b的变化对其造成的影响，在找到其最小值之后再去计算精度值。
* Q2：代价函数为什么选择二次方的形式？后面再说。
* Q3：如何找到代价函数的最小值？ -> 梯度下降（gradient descent）
    * >Summing up, the way the gradient descent algorithm works is to repeatedly compute the gradient $\nabla C$, and then to move in the opposite direction, "falling down" the slope of the valley.

    * 学习率（learning rate）的选择很重要，太大不下降，太小下降速度太慢。
    * ![image](https://cloud.githubusercontent.com/assets/5415646/25610167/c248ebd8-2f54-11e7-8d94-2c282822cbeb.png)
    * 梯度下降的替代方法是热门研究话题。

---

* 代价函数中对x求和，是对所有的训练样本求和，当训练样本很大时计算速度会较慢。解决方法 -> 随机梯度下降（stochastic gradient descent）。每次随机选取样本的一个子集（mini-batch），假设其梯度平均值等于所有样本的梯度平均值，即：$\frac{\sum_{j=1}^m \nabla C_{X_{j}}}{m} \approx \frac{\sum_x \nabla C_x}{n} = \nabla C$，然后仅对选取的子集计算梯度，可提高计算速度。

![image](https://cloud.githubusercontent.com/assets/5415646/25610187/d4fc11ba-2f54-11e7-97ef-846dfba69b18.png)

* 前面GD没有1/n，是因为C的定义里已经包含了1/n；而这里SGD除以m，是因为$\partial C_{X_j}$仅是C中的一项，没有除n，$X_j$一共有m项，故除以m。

---

* Q：这里的随机抽取有放回和无放回有无区别？经过一个epoch之后（抽取了n/m次），是否需要保证所有的样本都被抽取一次？还有无所谓？ -> 本章代码的做法是先将训练样本打乱，然后从头开始取出mini-batch个样本进行SGD，然后下一个mini-batch，直到左右样本都被取出一次，即完成一个epoch。

**Comment by hysic** [comment]
* 对于MNIST数据集，默认参数的SVM精度为94.35%，而经过fune-tune的SVM可以得到[98.56%的精度](http://peekaboo-vision.blogspot.jp/2010/09/mnist-for-ever.html)；两层神经网络得到的精度约为95%，而神经网络的最新记录为惊人的[99.79%](http://cs.nyu.edu/~wanli/dropc/)（这一结果要高于<http://yann.lecun.com/exdb/mnist/>上最好的0.23%错误率）。
    * sophisticated algorithm ≤ simple learning algorithm + good training data
